#STAGE 2. Domain composition 

###Software Requirements
[Interproscan]( https://www.ebi.ac.uk/interpro/interproscan.htm)

[hmmsearch](http://hmmer.org/)


1. Annotate "domain composition" of the input proteins (in our case sucy_database_uniprot.fasta) using interproscan and current realease of PfamA  (december 2016) 
```{r,eval=FALSE}
cd data
 
#old data
#/user/path/to/interproscan-5.4-47.0/interproscan.sh -appl TIGRFAM-13.0,ProDom-2006.1,PfamA-27.0,PrositeProfiles-20.97,SMART-6.2,HAMAP-201311.27,PrositePatterns-20.97,PRINTS-42.0,SuperFamily-1.75 -i sucy_database_uniprot.fasta -f tsv -pa -iprlookupD

#Run interproscan
/interproscan-5.21-60.0/interproscan.sh -appl PfamA-30.0 -i /input_data/sucy_database_uniprot.fasta -f tsv -pa -iprlookupD
```

2. Examinate the output example, using  following the output information of   [interproscan](https://github.com/ebi-pf-team/interproscan/wiki/InterProScan5OutputFormats)

```{r,eval=FALSE}
less -S sucy_database_uniprot.fasta.tsv
```


3. To get the markov models of the input  proteins, the following files are needed

```{r,eval=FALSE}
#1. Complete Pfam database of hmm 
#Get the current realease of Pfam database (heavy file, not provided!)
wget ftp://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam-A.hmm.gz
gunzip Pfam-A.hmm.gz
```

```{r,eval=FALSE}
#2. Especific domain identifiers of the input proteins 
#Using the output file from interproscan: sucy_database_uniprot.fasta.tsv,  we are interested in the 4° column:Analysis (e.g. Pfam / PRINTS / Gene3D) and the 5°: Signature Accession (e.g. PF09103 / G3DSA:2.40.50.140) 
cut -f 4,5 sucy_database_uniprot.fasta.tsv  > id_interpro.txt 
```

4. Obtain the specific markov models of the input proteins 
```{r,eval=FALSE}
#Run the script extract_hmms.pl . The latter use as input i) id_interpro.txt  and ii)  PfamA.hmm (Make sure you have those files in data/, and you are using those names, otherwise the script will not run )
perl ../scripts/extract_hmms.pl 
# Pfam
# hmms = 112 #pfam version 30. Using pfam version 27 we obtain a total of  114  hmms  
#This will generate an output file named 'my_Pfam.hmm'. (We We have compressed the file my_Pfam.hmm.bz2) 
#Note that Superfamily and TIGRFAM HMMs can also be used, but are commented out.
```

5. Annotate PFAM domains in the omic datasets using hmmsearch 

```{r,eval=FALSE}

#Genomic dataset 
cd data/
hmmsearch --cpu 8 --cut_ga -o /dev/null  --tblout genomes_refseq_nr_22122016.fa.pf.tab  my_Pfam.hmm genomic_dataset/genomes_refseq_nr_22122016.1.faa  & 
#Genomic dataset (each genome) 
#Optionally, compute the hmmsearch for each genome to calculate the SScore in Stage 4.Using the same command line  

#Genomic fragmented dataset 
nohup for i in /data/genomic_dataset_fragm/*.faa ; do hmmsearch --cpu 8 --cut_ga -o /dev/null --tblout $i.out.hmmsearch.tab my_Pfam.hmm  $i; done  & 

#Metagenomic dataset  

nohup for i in /data/metagenomic_dataset/* ; do hmmsearch --cpu 8 --cut_ga -o /dev/null --tblout $i.out.hmmsearch.tab my_Pfam.hmm  $i; done  & 
```

Check the output from the above command in the corresponding directories 

```{r,eval=FALSE}
ls *.tab data/genomic_dataset 
ls *.tab data/genomic_dataset_fragm 
ls *.tab data/metagenomic_dataset
```

